pkg/services/ngalert/store/image.go

```diff
@@ -11,84 +11,123 @@ import (
 	"github.com/grafana/grafana/pkg/services/sqlstore"
 )
 
+const (
+	imageExpirationDuration = 24 * time.Hour
+)
+
 type ImageStore interface {
-	// GetImage returns the image with the token or ErrImageNotFound.
+	// GetImage returns the image with the token. It returns ErrImageNotFound
+	// if the image has expired or if an image with the token does not exist.
 	GetImage(ctx context.Context, token string) (*models.Image, error)
 
-	// GetImages returns all images that match the tokens. If one or more
-	// tokens does not exist then it also returns ErrImageNotFound.
-	GetImages(ctx context.Context, tokens []string) ([]models.Image, error)
+	// GetImages returns all images that match the tokens. If one or more images
+	// have expired or do not exist then it also returns the unmatched tokens
+	// and an ErrImageNotFound error.
+	GetImages(ctx context.Context, tokens []string) ([]models.Image, []string, error)
 
 	// SaveImage saves the image or returns an error.
 	SaveImage(ctx context.Context, img *models.Image) error
 }
 
+type ImageAdminStore interface {
+	ImageStore
+
+	// DeleteExpiredImages deletes expired images. It returns the number of deleted images
+	// or an error.
+	DeleteExpiredImages(context.Context) (int64, error)
+}
+
 func (st DBstore) GetImage(ctx context.Context, token string) (*models.Image, error) {
-	var img models.Image
+	var image models.Image
 	if err := st.SQLStore.WithDbSession(ctx, func(sess *sqlstore.DBSession) error {
-		exists, err := sess.Where("token = ?", token).Get(&img)
+		exists, err := sess.Where("token = ? AND expires_at > ?", token, TimeNow().UTC()).Get(&image)
```

Comment:
1. I wonder if it would be useful if we distinguish between an image that does not exist in DB and an expired one by providing different types of errors. This may help us during the troubleshooting

Reply:
I started with this but later removed it as the tokens for `GetImages` can be either expired or not found, such that the caller of `GetImages` would want to know which tokens had expired and which tokens were not found. In the end I thought it would be better to make this pull request as simple as possible.